content/blog/eng/from-0-to-infinite-the-final-keyword.md

```diff
@@ -0,0 +1,338 @@
+
+In the previous example two main errors have been made, first error a Stack IS NOT an Array, second problem, is the cross-domain inheritance relationship. Simply, our domain classes must use implementations not to inherit them, in our example, stack was a concept of domain (Focusing only on reusing code can be a problem).
+
+You are probably thinking that the OOP is not keeping its promise. 
+However, let us remember that we are programming in OOP (Object Oriented Programming) not in OOI (Object Oriented Inheritance). There is always a trade off between code reuse and good design. 
+Inheritance should be mostly used for good design, for code reuse we go with composition.
+
+I invite you, however, to do some tests with your classes, to see if they adhere to these good practices and try to get a better design, avoiding inheritance.
+
+# The concrete example, break the immutable object 
+
+Let's see some example where the **final** keyword could be the right choice. 
+Through **final** we create a immutable object with minimal effort. This is only a **POC** to show how **final** works, there are many ways to resolve this problem. 
+
+Let's go see some code.
+
+```php
+class SomeImmutableObject
+{
+    private $someString;
+
+    public function __construct(string $value)
+    {
+        $this->someString = $value;
+    }
+
+    public function getValue(): string
+    {
+        return 'the value is:' . $this->someString . ' - ';
+    }
+}
+```
+
+Let's concentrate, can we modify this object after it has been created? …. Yes we can!
+```php
+$one = new SomeImmutableObject('Pippo');
+echo $one->getValue(); //Pippo
+$one->__construct('Pluto'); 
+echo $one->getValue(); //Pluto
+```
+
+try it! **[break the immutable object] (https://3v4l.org/CiHPG)**
+
+This problem is easy to solve, we must put a flag in the constructor and if it is true throw an exception. 
+_Another way to fix, is to create a `named constuctor` and make the `__constructor` private_. 
+let's do it with the first solution!
+
+# The concrete example, break the immutable object first fix
+
+```php
+<?php declare(strict_types=1);
+
+class SomeImmutableObject
+{
+    private $someString;
+    private $flagCreate = false;
+
+    public function __construct(string $value)
+    {
+        if (true === $this->flagCreate) {
+            throw new \BadMethodCallException('This is an immutable object has already create.');
+        }
+
+        $this->someString = $value;
+        $this->flagCreate = true;
+    }
+
+    public function getValue(): string
+    {
+        return 'the value is:' . $this->someString . ' - ';
+    }
+}
+
+$two = new SomeImmutableObject('Pippo');
+echo $two->getValue(); //Pippo
+$two->__construct('Pluto'); // \BadMethodCallException 
+echo $two->getValue();
+```
+
+Now we are happy and we have our immutable object, are we sure?, mmm no… look here! 
```

Comment:
Now we are happy and we have our immutable object. Are we sure? Mmm no… have a look at here!